File: baekjoon/특정한 최단 경로/1504.py
import sys

input = sys.stdin.readline

N,M = map(int, input().split(' '))
INF = float('INF')

distance = [[INF for _ in range(N + 1)] for _ in range(N + 1)]

for _ in range(M):
	start, end, weight = map(int, input().split(' '))
	distance[start][end] = weight
	distance[end][start] = weight

v1, v2 = map(int, input().split(' '))

for k in range(1, N + 1):
	for i in range(1, N + 1):
		for j in range(1, N + 1):
			if i == j:
				distance[i][j] = 0
			elif distance[i][j] > distance[i][k] + distance[k][j]:
				distance[i][j] = distance[i][k] + distance[k][j]

ans = min(distance[1][v1] + distance[v1][v2] + distance[v2][N], distance[1][v2] + distance[v2][v1] + distance[v1][N])

print(ans) if distance[1][v1] != INF or distance[v1][v2] != INF or distance[v2][N] != INF or distance[v1][N] != INF else print(-1)

File: baekjoon/특정한 최단 경로/1504.cpp
#include <iostream>
#include <algorithm>

using namespace std;

int dist[801][801] = {0, };

int main(){
	ios_base::sync_with_stdio(false);
	cin.tie(0);

	int N, M, INF = 987654321;
	int start, end, weight;
	cin >> N >> M;

	for (int i = 1; i <= N; i++){
		for (int j = 1; j <= N; j++){
			if (i != j) dist[i][j] = INF;
		}
	}

	while (M--){
		cin >> start >> end >> weight;

		dist[start][end] = weight;
		dist[end][start] = weight;
	}

	int v1,v2;
	cin >> v1 >> v2;

	for (int k = 1; k <= N; k++){
		for (int i = 1; i <= N; i++){
			for (int j = 1; j <= N; j++){
				if (i != j && dist[i][j] > dist[i][k] + dist[k][j]) dist[i][j] = dist[i][k] + dist[k][j];
			}
		}
	}

	int ans = min(dist[1][v1] + dist[v1][v2] + dist[v2][N], dist[1][v2] + dist[v2][v1] + dist[v1][N]);

	if (dist[1][v1] == INF || dist[v1][v2] == INF || dist[v1][N] == INF || dist[v2][N] == INF) cout << -1;
	else cout << ans;
}